### ISSUE TITLE
serialize the code in libzfs_mnttab_find

### ISSUE THREAD

**Issue by sandipd** [zfs-linux/zfs#29]
When multiple unmounts are ran in parallel, sometimes avl_find function call in libzfs_mnttab_find() returns false. False return value indicates the file system is not mounted. This behavior of avl_find is unpredicted, we do not have good solution to this problem. For now the following patch would serialize the code in libzfs_mnttab_find(). This needs to be removed in GA build.

Setting the target version to GA.

[anurag@localhost zfs-0.4.7]$ hg diff
diff -r 94fbf6c7a039 zfs-0.4.7/lib/libzfs/libzfs_dataset.c
--- a/zfs-0.4.7/lib/libzfs/libzfs_dataset.c    Tue Aug 10 20:22:58 2010 +0530
+++ b/zfs-0.4.7/lib/libzfs/libzfs_dataset.c    Tue Aug 10 20:44:14 2010 +0530
@@ -630,7 +630,9 @@
 {
     mnttab_node_t find;
     mnttab_node_t *mtn;
-    //int fd;
  +#if defined(LINUX_PORT)
-    int fd;
  +#endif
  
   if (!hdl->libzfs_mnttab_enable) {
       struct mnttab srch = { 0 };
  @@ -646,8 +648,10 @@
           return (ENOENT);
   }
-    //fd = open(FLOCK_PATH, O_RDONLY);
-    //flock(fd, LOCK_EX);
  +#if defined(LINUX_PORT)
-    fd = open(FLOCK_PATH, O_RDONLY);
-    flock(fd, LOCK_EX);
  +#endif
  
   if (avl_numnodes(&hdl->libzfs_mnttab_cache) == 0)
       libzfs_mnttab_update(hdl);
  @@ -656,10 +660,16 @@
   mtn = avl_find(&hdl->libzfs_mnttab_cache, &find, NULL);
   if (mtn) {
       *entry = mtn->mtn_mt;
-        //flock(fd, LOCK_UN);
  +#if defined(LINUX_PORT)
-        flock(fd, LOCK_UN);
-        close(fd);
  +#endif
       return (0);
   }
-    //flock(fd, LOCK_UN);
  +#if defined(LINUX_PORT)
-    flock(fd, LOCK_UN);
-    close(fd);
  +#endif
   return (ENOENT);
  }

[anurag@localhost ZFS_kq]$ hg heads -v
changeset:   79:13bce6630631
tag:         tip
user:        Prasad Joshi prasad@kqinfotech.com
date:        Tue Aug 10 20:50:03 2010 +0530
files:       zfs-0.4.7/lib/libzfs/libzfs_dataset.c
description:
Serialize the code in libzfs_mnttab_find(), this would avoid the problem of multiple umounts ran in parallel.

**Comment by sandipd** [comment]
Updated by Anand Mitra 2 months ago

Comment

This is a performance issue. The serialization will affect the performance. Address in GA
#2
Updated by Anand Mitra 14 days ago

```
* Target version changed from GA to Parked
```

Comment
- parked this issue serialization in the userland mount code is not a big issue.

**Comment by mitra** [comment]
working as desired
